### PR TITLE
[None][fix] clamp very small non-zero sampling temperature

### LINKED ISSUE #15715: [Bug]: Clamp very small non-zero temperature values to avoid numerical instability
### System Info

CPU architecture: x86_64

CPU: Intel Xeon Gold 6462C, 8 vCPUs (4 cores × 2 threads), max 3900 MHz

GPU name: NVIDIA L20
GPU memory size: 48G

TensorRT-LLM branch or tag: main
TensorRT-LLM commit: https://github.com/NVIDIA/TensorRT-LLM/commit/28acbad14b442e89312cfe9e845a5f721a3e702f
PyTorch: 2.10.0+cu130
CUDA: 13.0.88
cuDNN: 9.15.0
NVIDIA driver version: 550.163.01

OS: Ubuntu 24.04.3 LTS

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

```python
from tensorrt_llm.sampling_params import SamplingParams

# Before the fix: very small non-zero temperature bypasses validation
# and enters the sampling backend, causing potential inf/nan in logits/temperature
sp = SamplingParams(temperature=1e-12, top_k=50)
print(sp.temperature)  # prints 1e-12 — dangerously small

# Simulating what happens inside the backend:
import torch
logits = torch.tensor([1.0, 2.0, 3.0])
temperature = 1e-12
scaled = logits / temperature
print(scaled)  # tensor([1.0e+12, 2.0e+12, 3.0e+12]) → leads to inf/nan in softmax
print(torch.softmax(scaled, dim=-1))  # tensor([0., 0., 1.]) or nan depending on dtype
```

### Expected behavior

SamplingParams should reject or safely normalize temperature values that are 
positive but below a numerical stability threshold.

Specifically:
- temperature=0.0  → kept as 0.0 (greedy decoding)
- temperature=1e-12 → clamped to MIN_SAMPLING_TEMPERATURE (1e-2)
- temperature=1.0  → kept as 1.0 (normal sampling)

### actual behavior

amplingParams only validates temperature >= 0, so values like 1e-12 
pass validation unchanged and are forwarded to the sampling backend.

Inside the backend, logits / temperature with temperature=1e-12 produces 
extremely large logit values (e.g., 1e+12), which overflow to inf or nan 
in float16/bfloat16 precision, leading to undefined sampling behavior.

Example:
```python
import torch
logits = torch.tensor([1.0, 2.0, 3.0], dtype=torch.float16)
scaled = logits / 1e-12
print(scaled)
# tensor([inf, inf, inf], dtype=torch.float16)

print(torch.softmax(scaled, dim=-1))
# tensor([nan, nan, nan], dtype=torch.float16)
```

### additional notes

1.  This fix mirrors the same numerical stability guard used by vLLM    (temperature = max(temperature, 1e-2) for non-zero values).

2. The issue is most severe in float16/bfloat16: typical logit values (~10)    divided by 1e-6 already exceed float16's max range (~65504),    producing inf/nan in softmax.

3. temperature=0.0 (greedy decoding) is intentionally excluded from the clamp    and its behavior is unchanged.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by lonexreb**
## Description

Fixes #15715.

`SamplingParams` only validated `temperature >= 0`, so tiny positive values (e.g. `1e-12`) passed through unchanged. In the sampling backend, `logits / temperature` with such values overflows to `inf`/`nan` in fp16/bf16 (typical logits ~10 divided by 1e-6 already exceed fp16's max ~65504), producing undefined sampling behavior.

## Change

- Add `MIN_SAMPLING_TEMPERATURE = 1e-2` and clamp any positive temperature below it in `SamplingParams._validate()`.
- `temperature == 0.0` (greedy decoding) is explicitly left unchanged.
- Mirrors vLLM's numerical-stability guard.

## Test

New CPU-only unit tests in `tests/unittest/llmapi/test_sampling_params.py` covering: `None` passthrough, greedy `0.0` preserved, tiny values (`1e-12`, `1e-6`, `5e-3`) clamped to the floor, normal temperatures unchanged, and negative rejected.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved sampling stability by clamping extremely small positive temperature values to a safe minimum.
  * Preserved greedy decoding when temperature is set to zero.
  * Continued rejecting negative temperature values.

* **Tests**
  * Added coverage for default, boundary, valid, and invalid temperature settings.

### CODE REVIEW

**Comment by lonexreb** [comment]
Thanks for the review. Addressed in ef4292256:
- **Non-finite temperatures**: now reject `NaN`/`±Inf` with a `ValueError` before the sign/clamp checks (they previously bypassed both comparisons).
- **Type annotations**: added `-> None` and `float` param annotations to the new tests.
- **Test assertions**: assert the contractual floor `1e-2` directly (not the constant), and added NaN/Inf rejection tests.

**Comment by lori-ren** [comment]
Closing in favor of #15716
